src/wp-admin/includes/class-wp-plugin-dependencies.php

```diff
@@ -0,0 +1,634 @@
+	}
+
+	/**
+	 * Loop installed plugins and process dependencies.
+	 *
+	 * @since 5.9.0
+	 * @access public
+	 *
+	 * @return void
+	 */
+	public function loop_installed_plugins() {
+		foreach ( $this->installed_plugins as $file => $plugin ) {
+			$this->maybe_process_plugin_dependencies( $file );
+		}
+	}
+
+	/**
+	 * Check plugin dependencies.
+	 *
+	 * @since 5.9.0
+	 * @access public
+	 *
+	 * @param string $file The plugin file.
+	 *
+	 * @return void
+	 */
+	public function maybe_process_plugin_dependencies( $file ) {
+
+		$plugin_is_active           = is_plugin_active( $file );
+		$plugin_awaiting_activation = in_array( $file, $this->get_plugins_to_activate(), true );
+
+		// Early return if the plugin is not active or we don't want to activate it.
+		if ( ! $plugin_is_active && ! $plugin_awaiting_activation ) {
+			return;
+		}
+
+		// Get the dependencies.
+		$dependencies = $this->get_plugin_dependencies( $file );
+
+		// Early return if there are no dependencies.
+		if ( empty( $dependencies ) ) {
+			return;
+		}
+
+		// Loop dependencies.
+		$dependencies_met = true;
+		foreach ( $dependencies as $dependency ) {
+
+			// Set $dependencies_met to false if one of the dependencies is not met.
+			if ( ! $this->process_plugin_dependency( $file, $dependency ) ) {
+				$dependencies_met = false;
+			}
+		}
+
+		$in_circular_dependency = $this->in_circular_dependency( $file );
+
+		if ( ! $dependencies_met ) {
+
+			// Make sure plugin is deactivated when its dependencies are not met.
+			if ( $plugin_is_active && ! $in_circular_dependency ) {
+				deactivate_plugins( $file );
+			}
+
+			// Add plugin to queue of plugins to be activated.
+			$this->add_plugin_to_queue( $file );
+
+		} elseif ( $plugin_awaiting_activation ) {
+			activate_plugin( $file );
+			$this->remove_plugin_from_queue( $file );
+		}
+	}
+
+	/**
+	 * Get an array of dependencies.
+	 *
+	 * @since 5.9.0
+	 * @access public
+	 *
+	 * @param string $file The plugin file.
+	 *
+	 * @return array
+	 */
+	public function get_plugin_dependencies( $file ) {
+		if ( ! isset( $this->plugin_dependencies[ $file ] ) ) {
+			$this->plugin_dependencies[ $file ] = array();
+			$plugin_dependencies = get_plugin_data( WP_PLUGIN_DIR . '/' . $file )['RequiresPlugins'];
+			if ( empty( $plugin_dependencies ) ) {
+				$this->plugin_dependencies[ $file ] = array();
+				return array();
+			}
+
+			$plugin_dependencies = str_getcsv( $plugin_dependencies );
+			foreach ( $plugin_dependencies as $key => $dependency ) {
+				$this->plugin_dependencies[ $file ][] = array( 'slug' => trim( $dependency ) );
+			}
+		}
+
+		return $this->plugin_dependencies[ $file ];
+	}
+
+	/**
+	 * Processes a plugin dependency.
+	 *
+	 * @since 5.9.0
+	 * @access protected
+	 *
+	 * @param string   $plugin     The plugin defining the dependency.
+	 * @param stdClass $dependency A dependency.
+	 *
+	 * @return bool
+	 */
+	protected function process_plugin_dependency( $plugin, $dependency ) {
```

Comment:
One thing I dončt personally like about this method is that it does too many things. It will process the plugin dependency (which the name suggests), but also returns a value (something I wouldn't expect out of this method).

I'm not 100% sure what the false/true means in this context?

Reply:
I added a comment to explain what the return value is. It returns false when the plugin has unmet dependencies and true if all dependencies are met (or there are no dependencies - in which case there are no unmet dependencies).

Reply:
Added a new `has_unmet_dependencies` method and removed the return from this one. There is now proper separation :+1: